scopes.md

```diff
@@ -0,0 +1,138 @@
+- Post/remove likes
+- Subscribe/mute posts
+- Hide posts from streams
+
+### openid
+
+This scope is defined by the OpenID Connect spec. This is a mandatory scope. It gives access to OpenID Connect authentication.
+
+### private
+
+This scope gives access to the following requests:
+- Read private posts
+- Read private posts intercations (likes, comments)
+
+Reading private posts of contacts additionaly requires `contacts:read` scope.
+
+### public
+
+This scope gives access to the following requests:
+- Read public profiles of users
+- Perform search among publicly searchable profiles
+- Read public posts
+- Read reshares
+- Read public posts intercations (likes, comments)
+
+### posts:publish
+
+This scope gives access to the following requests:
+- Create posts
+- Create reshares
+- Post photos
+
+Create private posts additionaly requires `contacts:read` scope.
+
+### posts:modify
+
+This scope gives access to the following requests:
+- Delete posts
+- Delete photos
+
+### profile
+
+This scope is defined by the OpenID Connect spec. It allows reading user claims (user properties). Claims are read using UserInfo route or from ID Token.
+
+User claims defined by the OpenID Connect are somewhat similar to users own profile data. Generally this scope means "Allow reading user's own profile". Besides user claims defined by the standard this scope also gives access to "read own profile" diaspora* API endpoint.
+
+Supported claims by diaspora:
+
+| claim      | meaning                                             |
+| ---------- | --------------------------------------------------- |
+| name       | Full name                                           |
+| nickname   | The left part of diaspora handle before `@` sign    |
+| profile    | User's profile URL address                          |
+| picture    | User avatar URL address                             |
+| gender     | Gender profile field                                |
+| birthdate  | Birth date profile field                            |
+| locale     | UI language selection specified in account settings |
+| updated_at | Date of last profile edit                           |
+
+### profile:modify
+
+This scope gives access to updating users own profile
+
+### tags
+
+This scope gives access to read/create/delete tag followings.
+
+### streams
+
+This scope gives access streams.
+
+There are 3 types of streams available. Accessing each stream requires two scopes: `streams` scope and a scope specific for the stream type.
+
+| Stream name     | Stream specifics                           | Access scopes               |
+| --------------- | ------------------------------------------ | --------------------------- |
+| Activity stream | Includes posts with users own activity     | `streams` + `interactions`  |
+| Aspects stream  | Includes posts from specific aspects       | `streams` + `contacts:read` |
+| Tags streams    | Includes posts which contain specific tags | `streams` + `tags`          |
```

Comment:
This makes me wonder whether we should actually split "tags" into a read and a write scope.

Reply:
While I don't feel like it makes sense to create a specific scope for just reading followed tag list, I don't see a better way to make it consistent, so I'm fine with splitting the `tags` scope in `tags:read` and `tags:modify`. Or maybe it's better to make `tags:read` and just `tags` where `tags:read` is for reading and `tags` is for full access?

Reply:
`tags:read` and `tags:modify` would follow the pattern we have so far, so 👍 from my side but I'd like to hear more opinions on this :)

Reply:
The potential benefit of using permissions could be that clients automatically get what they can see if we allow a user to update an app's permissions (check/uncheck private for example).

Reply:
> The potential benefit of using permissions could be that clients automatically get what they can see if we allow a user to update an app's permissions (check/uncheck private for example).

You mean having a global public/private versus a streams:tags, streams:activity, etc?  They could just flag off public/private globally instead?

Reply:
Oh sorry I think I misunderstood you, I thought you were talking about endpoints returning subsets rather than creating more scopes, disregard then, I have no strong opinion on either. This whole scope combination thing makes it harder to follow what to request, but a ton of scopes also doesn't help finding a good overview... so I dunno

Reply:
@jhass streams are the one area that I can see doing the combination.  "I already told you you can read my tags so give me the tags stream."  Some of the other combinations I am more concerned about.  I'm not sure if we even want people to be able to say "only give me limited posts by other users that shared it with me".  That seems overly fine grained, and it's not something users do through the current UI AFAIK.  However doing that by combining the private flag and the contacts:read flag seems like a recipe for people being very confused about what they are authorizing.

Reply:
> I'm not sure if we even want people to be able to say "only give me limited posts by other users that shared it with me". That seems overly fine grained, and it's not something users do through the current UI AFAIK. However doing that by combining the private flag and the contacts:read flag seems like a recipe for people being very confused about what they are authorizing.

@HankG, I can see your point. What I think is that if we show other people private posts means that you kinda can know who your contacts are, which overlaps with the contacts access scope. That is why I though of dividing scopes this way. However I feel now that it probably makes few sense to have specific access control for own data and other people data. I think that we can just use `private` scope for all private posts and don't require `contacts:read` for this. The only thing is that we should communicate this well in our UI, that your API client can know who share with you when you give access to private posts.

However we still need `contacts:read` for the aspects stream, because aspects stream supposes that you know what the aspects are and this is controlled by the `contacts:read` scope.

Reply:
Maybe we just combine the scopes which I defined as `contacts:read` and `private` into one? Let's call it for instance `aspects:read`. It'll allow:

> - Read aspect list
> - Read aspects membership
> - Read private user profiles of contacts
> - Read contacts of contacts when allowed by contact's aspect setting
> - User search will include contacts who are hidden from public search
> - Read private posts
> - Read private posts interactions (likes, comments)

This will cover the fact that reading private posts always means that you know your contacts. Also  by having the compound scope we can add "Reading aspects stream" to the list of permissions and thus
1. make it easier to track what is being permitted
2. avoid the need for a specific scope for streams at all

After doing so we can throw away the `streams` scope at all and use `aspects:read` as the only requirement for aspects stream and interactions streams.

Reply:
Another way we could do the same thing is to introduce "scope dependencies". For instance one can allow `contacts:read` separately, but if you request the scope for private posts access, this would require granting `contacts:read` too

Reply:
> Read contacts of contacts when allowed by contact's aspect setting

I don't think this is supposed to be allowed any more.  I can't see a way to do it in the UI and I don't see a way for a user to disable it through the UI any more.

Reply:
> Read private user profiles of contacts
This is really a decision by the other user to allow others to see their private profile data or not and is driven by whether they have them in their contact list or not.  Would this be to add an additional control on top of that to block performing that query regardless?

Reply:
I think it may be better to have this as an interactive discussion and then summarize it here somewhere...

Reply:
> I don't think this is supposed to be allowed any more. I can't see a way to do it in the UI and I don't see a way for a user to disable it through the UI any more.

This is correct.

> This is really a decision by the other user to allow others to see their private profile data or not and is driven by whether they have them in their contact list or not. Would this be to add an additional control on top of that to block performing that query regardless?

These are different levels. We're talking here about API access scopes, which means the access that is given by a diaspora user to a API client application. Diaspora users may share their profile privately with others, but whether users wants to allow their 3rd party API client to access their private contacts is a whole different story. I may trust a 3rd party application my public profile data but at the same time not trust my contacts data. And it is a different thing. We need two different levels of access control for this.

Reply:
I'm not sure if we **need it**.  We may decide to offer that sort of thing.

Reply:
I think the extra controls on the last three bullets are superfluous.  I don't understand the usage of them at all.

Reply:
> I think the extra controls on the last three bullets are superfluous. I don't understand the practical usage of them at all. If it seems that's needed it can be added later if we get feedback about needing finer grained controls. Adding it right now doesn't seem appropriate.

But why? Do you think that every possible API client should have access to private publications? Let's say I just want to use diaspora API for authorization to a 3rd party website with diaspora account. Why would I want to also allow this 3rd party website to read private posts shared with me?

Reply:
I understand why you'd want to limit access to that entire data set, so you wouldn't grant read privileges to that endpoint.  Within the endpoint it seems like an artificial differentiation.  Most users wouldn't even understand the difference in a private post versus a public post in the context of setting those permissions.  It gets even hairier with comments and likes on that post.  I'm not opposed to it entirely if it is done in a way that doesn't create confusion with stacked permissions.

Reply:
I think it should be possible to prevent reading private posts, because maybe you want to grant something permissions to create posts automatically (and if you want it to create private posts, it needs to be able to read aspects, so that needs to be a different permission), but I don't want something that should only create posts to be able to read private posts, so it needs to be different permissions.